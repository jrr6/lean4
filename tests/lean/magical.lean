-- Projection to the witness should be rejected.
def witness : Nat := (⟨1, Nat.le_refl _⟩ : ∃ x, x ≥ 1).1

-- Projection to the property as well (it could contain the witness projection).
theorem witness_eq (h : ∃ x : Nat, True) : h.2 = h.2 := rfl

inductive P : Nat → n > 0 → Prop
  | mk (n) (q : n > 0) : P n q

def foo (h1 : P n h) (h2 : P m h') := h1.1 = h2.2
def test (h : Nat) := h.2

example (a : Nat) (h : ∃ x : Nat, x > a + 1) : ∃ x : Nat, x > a := by
  cases h with
  | intro w hw =>
    exists w
    omega
