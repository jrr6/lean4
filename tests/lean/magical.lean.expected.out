magical.lean:2:21-2:56: error: Invalid projection: Cannot project a value of non-propositional type
  Nat
from the expression
  Exists.intro 1 (Nat.le_refl 1)
which has propositional type
  ∃ x, x ≥ 1
magical.lean:5:8-5:18: error: (kernel) invalid projection
  h.2
magical.lean:10:45-10:49: error: Invalid projection: Index `2` is invalid for this structure; the only valid index is 1

Note: The expression
  h2
has type
  P m h'
which has only 1 field
magical.lean:11:22-11:25: error: Invalid projection: Expected a value whose type is a structure
  h
has type
  Nat
